### PR TITLE
Added customTargets and example project to test.

### PR DESCRIPTION
**Pull request by Jmales**

### CODE REVIEW

File: TestBundle/TestBundle.csproj [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+   <!-- Defines Info.plist of ServiceStudio.app -->
+   <PropertyGroup>
+    <CFBundleName>TestBundle</CFBundleName> <!-- Also defines .app file name -->
+    <CFBundleDisplayName>TestBundle</CFBundleDisplayName>
+    <CFBundleShortVersionString>TestBundle</CFBundleShortVersionString>
+    <CFBundleIconFile>IconTest.icns</CFBundleIconFile> <!-- Will be copied from output directory -->
+    <NSPrincipalClass>NSApplication</NSPrincipalClass>
+    <CFBundlePackageType>AAPL</CFBundlePackageType>
+    <CFBundleSignature>4242</CFBundleSignature>
+    <NSRequiresAquaSystemAppearance>true</NSRequiresAquaSystemAppearance> <!-- TEMP: Remove black border from window when on dark theme -->
```
Comment (joaompneves) [inline TestBundle/TestBundle.csproj]: remove the comment

File: TestBundle/TestBundle.csproj [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+              <command>dotnet pack ..\\Dotnet.Bundle\\ -p:Configuration=$(Configuration) -p:PackageVersion=$(DotnetBundleTestVersion)</command>
+            </Command>
+            <Command>
+              <type>AfterBuild</type>
+              <command>dotnet msbuild -t:BundleApp -p:RuntimeIdentifier=osx-x64 -p:Configuration=$(Configuration)</command>
+            </Command>
+          </CustomCommands>
+        </CustomCommands>
+  </PropertyGroup>
+   <!-- Defines Info.plist of ServiceStudio.app -->
```
Comment (joaompneves) [inline TestBundle/TestBundle.csproj]: remove

File: TestBundle/TestBundle.csproj [hunk 3]
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk">
+	<PropertyGroup>
+		<GenerateAssemblyInfo>false</GenerateAssemblyInfo>
```
Comment (joaompneves) [inline TestBundle/TestBundle.csproj]: fix identation

File: DotNet.Bundle/PlistWriter.cs [hunk 4]
```diff
@@ -91,6 +102,45 @@ private void WriteProperty(XmlWriter xmlWriter, string name, bool value)
...
+                xmlWriter.WriteStartElement("string");
+                xmlWriter.WriteString(value.GetMetadata(CFBundleURLName));
+                xmlWriter.WriteEndElement();
+
+                xmlWriter.WriteStartElement("key");
+                xmlWriter.WriteString(CFBundleURLSchemes);
+                xmlWriter.WriteEndElement();
+
+                xmlWriter.WriteStartElement("array");
+                string[] urlSchemes = value.GetMetadata(CFBundleURLSchemes).Split(Separator);
```
Comment (joaompneves) [inline DotNet.Bundle/PlistWriter.cs]: var

File: DotNet.Bundle/PlistWriter.cs [hunk 5]
```diff
@@ -91,6 +102,45 @@ private void WriteProperty(XmlWriter xmlWriter, string name, bool value)
...
+            xmlWriter.WriteStartElement("key");
+            xmlWriter.WriteString(name);
+            xmlWriter.WriteEndElement();
+
+            xmlWriter.WriteStartElement("array");
+            xmlWriter.WriteStartElement("dict");
+
+            foreach (var value in values)
+            {
+                xmlWriter.WriteStartElement("key");
```
Comment (joaompneves) [inline DotNet.Bundle/PlistWriter.cs]: WriteProperty(xmlWriter, CFBundleURLName, CFBundleURLName)

File: README.md [hunk 6]
```diff
@@ -34,7 +34,22 @@ Define properties to override default bundle values
...
+
+<ItemGroup>
+    <!-- Optional URLTypes.Check TestBundle.csproj for a working example. -->
+    <CFBundleURLTypes Include="dummy"> <!-- The name of this file is irrelevant, it's a MSBuild requirement.-->
+        <CFBundleURLName>TestApp URL</CFBundleURLName>
+        <CFBundleURLSchemes>testappurl;testappurl://</CFBundleURLSchemes> <!-- Note the ";" separator-->
+    </CFBundleURLTypes>
+    <CFBundleURLTypes Include="dummy">
+        <CFBundleURLName>TestApp URL2</CFBundleURLName>
+        <CFBundleURLSchemes>test://;</CFBundleURLSchemes> <!-- Be sure to include a ";" separator as well when having 1 parameter, since CFBundleURLSchemes makes it mandatory to be an array -->
```
Comment (joaompneves) [inline README.md]: it shouldnt be mandatory... you should handle that

File: DotNet.Bundle/PlistWriter.cs [hunk 7]
```diff
@@ -86,6 +101,61 @@ private void WriteProperty(XmlWriter xmlWriter, string name, bool value)
...
+            xmlWriter.WriteEndElement();
+
+            xmlWriter.WriteStartElement("array");            
+
+            foreach (var value in values)
+            {
+                xmlWriter.WriteStartElement("dict");
+
+                var metadataDictionary = value.CloneCustomMetadata();
+                foreach (var key in metadataDictionary.Keys)
```
Comment (joaompneves) [inline DotNet.Bundle/PlistWriter.cs]: foreach (var entry in metadataDictionary)

File: DotNet.Bundle/PlistWriter.cs [hunk 8]
```diff
@@ -1,13 +1,16 @@
...
+
 namespace Dotnet.Bundle
 {
     public class PlistWriter
     {
         private readonly BundleAppTask _task;
-        private readonly StructureBuilder _builder;
+        private readonly StructureBuilder _builder;
+
+        private const char Separator = ';';
```
Comment (joaompneves) [inline DotNet.Bundle/PlistWriter.cs:16]: private const string Separator = ";";

File: DotNet.Bundle/PlistWriter.cs [hunk 9]
```diff
@@ -86,6 +101,61 @@ private void WriteProperty(XmlWriter xmlWriter, string name, bool value)
...
+            foreach (var value in values)
+            {
+                xmlWriter.WriteStartElement("dict");
+
+                var metadataDictionary = value.CloneCustomMetadata();
+                foreach (var key in metadataDictionary.Keys)
+                {
+                    var dictValue = metadataDictionary[key].ToString();
+
+                    if (dictValue.Contains(Separator.ToString())) //array
```
Comment (joaompneves) [inline DotNet.Bundle/PlistWriter.cs]: the if seems unecessary

File: README.md (hunk 6 above)
Comment (Jmales) [reply README.md]: In CFBundleURLSchemes if you have 1 value, you need the structure 
`<array>
     <string>value</string>
 </array>` (https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html#//apple_ref/doc/uid/20001431-102207).

But that's not the default behavior, and I want to make this agnostic for all metadata's parameters that we writing to xml... 🤔

File: DotNet.Bundle/PlistWriter.cs (hunk 8 above)
Comment (Jmales) [reply DotNet.Bundle/PlistWriter.cs:16]: The split takes a character, hence the `const char` that  can use in both the Split() and Contains() methods

File: README.md (hunk 6 above)
Comment (joaompneves) [reply README.md]: but you can add the ";" if doesnt have... I mean this implementation detail shouldnt be exposed to the developer

Comment (Jmales) [reply README.md]: Hmm ok, I see what you mean. Ty

File: DotNet.Bundle/PlistWriter.cs [hunk 10]
```diff
@@ -1,13 +1,17 @@
...
 namespace Dotnet.Bundle
 {
     public class PlistWriter
     {
         private readonly BundleAppTask _task;
-        private readonly StructureBuilder _builder;
+        private readonly StructureBuilder _builder;
+
+        private const char Separator = ';';
+        private readonly string[] PropertiesWithMandatoryArray = { "CFBundleURLSchemes" };
```
Comment (joaompneves) [inline DotNet.Bundle/PlistWriter.cs]: private static readonly ArrayTypeProperties

File: DotNet.Bundle/PlistWriter.cs [hunk 11]
```diff
@@ -86,6 +102,63 @@ private void WriteProperty(XmlWriter xmlWriter, string name, bool value)
...
+            xmlWriter.WriteEndElement();
+
+            xmlWriter.WriteStartElement("array");            
+
+            foreach (var value in values)
+            {
+                xmlWriter.WriteStartElement("dict");
+                var metadataDictionary = value.CloneCustomMetadata();
+
+                foreach (var entry in metadataDictionary.Keys)
```
Comment (joaompneves) [inline DotNet.Bundle/PlistWriter.cs]: foreach (var entry in metadataDictionary)
{
    var key = entry.Key.ToString();
    var value = entry.Value.ToString();

    if (value.Contains(Separator.ToString()) || PropertiesWithMandatoryArray.Contains(entry))
    {
        WriteProperty(xmlWriter, entryString, value.Split(Separator));
    } else {
        WriteProperty(xmlWriter, entryString, value);
    }
}

Comment (Jmales) [reply DotNet.Bundle/PlistWriter.cs]: tried that first, but doesn't work.
image

Comment (joaompneves) [reply DotNet.Bundle/PlistWriter.cs]: foreach (DictionaryEntry entry in metadataDictionary)

File: DotNet.Bundle/PlistWriter.cs [hunk 12]
```diff
@@ -11,7 +12,7 @@ public class PlistWriter
         private readonly StructureBuilder _builder;
 
         private const char Separator = ';';
-        private readonly string[] PropertiesWithMandatoryArray = { "CFBundleURLSchemes" };
+        private static readonly List<string> ArrayTypeProperties = new List<string>{ "CFBundleURLSchemes" };
```
Comment (joaompneves) [inline DotNet.Bundle/PlistWriter.cs]: dont need a list use Linq!

File: DotNet.Bundle/PlistWriter.cs [hunk 13]
```diff
@@ -86,6 +104,62 @@ private void WriteProperty(XmlWriter xmlWriter, string name, bool value)
...
+
+                foreach (DictionaryEntry entry in metadataDictionary)
+                {
+                    var dictValue = entry.Value.ToString();
+                    var dictKey = entry.Key.ToString();
+
+                    if (dictValue.Contains(Separator.ToString()) || ArrayTypeProperties.Contains(dictKey)) //array
+                    {
+                        WriteProperty(xmlWriter, dictKey, dictValue.Split(Separator));
+                    } else {
```
Comment (joaompneves) [inline DotNet.Bundle/PlistWriter.cs]: fix { identation to match project style

**Review by joaompneves — APPROVED** [review]

File: DotNet.Bundle/PlistWriter.cs (hunk 13 above)
Comment (Jmales) [reply DotNet.Bundle/PlistWriter.cs]: nice catch :P

**Review by joaompneves — APPROVED** [review]

**Pull request merged**